include/irrArray.h

```diff
@@ -131,106 +99,19 @@ class array
 	\param index: Where position to insert the new element. */
 	void insert(const T& element, u32 index=0)
 	{
-		_IRR_DEBUG_BREAK_IF(index>used) // access violation
-
-		if (used + 1 > allocated)
-		{
-			// this doesn't work if the element is in the same
-			// array. So we'll copy the element first to be sure
-			// we'll get no data corruption
-			const T e(element);
-
-			// increase data block
-			u32 newAlloc;
-			switch ( strategy )
-			{
-				case ALLOC_STRATEGY_DOUBLE:
-					newAlloc = used + 5 + (allocated < 500 ? used : used >> 2);
-					break;
-				default:
-				case ALLOC_STRATEGY_SAFE:
-					newAlloc = used + 1;
-					break;
-			}
-			reallocate( newAlloc);
-
-			// move array content and construct new element
-			// first move end one up
-			for (u32 i=used; i>index; --i)
-			{
-				if (i<used)
-					allocator.destruct(&data[i]);
-				allocator.construct(&data[i], data[i-1]); // data[i] = data[i-1];
-			}
-			// then add new element
-			if (used > index)
-				allocator.destruct(&data[index]);
-			allocator.construct(&data[index], e); // data[index] = e;
-		}
-		else
-		{
-			// element inserted not at end
-			if ( used > index )
-			{
-				// create one new element at the end
-				allocator.construct(&data[used], data[used-1]);
-
-				// move the rest of the array content
-				for (u32 i=used-1; i>index; --i)
-				{
-					data[i] = data[i-1];
-				}
-				// insert the new element
-				data[index] = element;
-			}
-			else
-			{
-				// insert the new element to the end
-				allocator.construct(&data[index], element);
-			}
-		}
-		// set to false as we don't know if we have the comparison operators
+		_IRR_DEBUG_BREAK_IF(index > m_data.size()) // access violation
+		auto pos = std::next(m_data.begin(), index);
+		m_data.insert(pos, element);
 		is_sorted = false;
-		++used;
 	}
 
-
 	//! Clears the array and deletes all allocated memory.
 	void clear()
 	{
-		if (free_when_destroyed)
-		{
-			for (u32 i=0; i<used; ++i)
-				allocator.destruct(&data[i]);
-
-			allocator.deallocate(data); // delete [] data;
-		}
-		data = 0;
-		used = 0;
-		allocated = 0;
+		std::vector<T>().swap(m_data);

```

Comment:
`m_data.clear()`

Reply:
I was conflicted about this one, because the comment says "Clears the array and deletes all allocated memory.". Generally clear() does not delete the allocated memory, it only resets the size to 0.
clear() followed by shrink_to_fit() would be more like it (although the spec allows the library to ignore shrink_to_fit).
This in-place swap is the usual way to guarantee a total free.

I will look at how array::clear() is used and see if there are any situations that rely on this behavior.

Reply:
In that case just add a comment so it's obvious why the code is that way :)

Reply:
added comment